File: 209/minimum-size-subarray-sum.js
/**
 * @param {number} s
 * @param {number[]} nums
 * @return {number}
 */
var minSubArrayLen = function (s, nums) {
    let len = nums.length;
    let minLen = Infinity;
    let start, end;
    for (start = 0; start < len; start++) {
        let sum = nums[start];
        for (end = start + 1; end <= len; end++) {
            if (sum < s) {
                sum += nums[end];
                continue;
            } else {
                minLen = Math.min(minLen, end - start);
            }
        }
    }
    return minLen == Infinity ? 0 : minLen;
};

File: 209/MinimumSizeSubArraySum.scala
object MinimumSizeSubArraySum {
  def minSubArrayLen(s: Int, nums: Array[Int]): Int = {
    val len = nums.length

    @annotation.tailrec
    def findMinLen(low: Int, high: Int, sum: Int, minLen: Int): Int = {
      if (high >= len) {
        if (low < len && sum - nums(low) >= s) findMinLen(low+1, high, sum - nums(low), minLen min (high - low))
        else if (minLen != Int.MaxValue || sum >= s) minLen min (high - low)
        else 0
      }
      else if (sum < s) findMinLen(low, high + 1, sum + nums(high), minLen)
      else findMinLen(low+1, high, sum - nums(low), minLen min (high - low))
    }

    findMinLen(0, 0, 0, Int.MaxValue)
  }

  def main(args: Array[String]): Unit = {
    val s = 7
    val nums = Array(2,3,1,2,4,3)
    println(minSubArrayLen(s, nums))
  }
}
